README.md
prc library
===========

prc implements a few functions to compute
[Precision-Recall](https://en.wikipedia.org/wiki/Precision_and_recall)
curves, and their AUC (area under curve). The implementation is
checked against [scikit-learn](https://scikit-learn.org/).

## Installation

Latest release
```
opam install prc
```

Latest commit
```
opam pin add -y prc https://github.com/pveber/prc.git
```

## Motivation

Precision-Recall curves are a useful representation of the performance
of binary classification methods in the case there are many negative
items. Problem is they are difficult to estimate from a finite
sample. As an illustration, compare the following graph obtained in
a particular case where the true precision-recall curve can be
computed analytically (so-called binormal model):

![Empirical VS true precision-recall curves](img/true_vs_empirical_rp_curves.png)

The red curve shows the "true" precision-recall curve while the grey
ones are the empirical estimated obtained from samples of
size 1000. We see that there is a lot of variability, particularly in
the low-recall region.

As a consequence, displaying a precision-recall curve can be
misleading, and it is safer to report an estimate of its area under
curve (AUC) along with a confidence interval. This is what this
library provides.

## Quick test

Demo code is available in the `demo` directory. To run it, simply
launch an interpreter via a `dune utop demo` command, and then call
functions in `Prc_demo`. For instance to get an overview of the
sampling distribution of several estimators under the binormal model,
just type:
```ocaml
# Prc_demo.estimator_sampling_distribution ~sample_size:1000 ();;
```

This yields:

![demo](img/estimators_sampling_distribution.png)

where each boxplot represents the distribution of an estimator
(resp. binormal, trapezoidal and average precision) for a binormal
model when drawing finite samples of size 1000.


CHANGES
v0.1.0 2021-11-02 Villeurbanne
------------------------------

- documentation improvements
- added bootstrap procedure to compute confidence intervals
- added demo code to check the procedures perform ok.

v0.0.0 2021-10-08 Maillezais
----------------------------

First release on OPAM


test/prc_test.ml
open Core_kernel
open Prc

let floats_are_close_enough x y = Float.(abs (x -. y) < 1e-6)

module Naive_implementation = struct
  let tp (Dataset xs) c = List.count xs ~f:(fun (x, b) -> b && Float.(x >= c))
  let _tn (Dataset xs) c = List.count xs ~f:(fun (x, b) -> not b && Float.(x < c))
  let fp (Dataset xs) c = List.count xs ~f:(fun (x, b) -> not b && Float.(x >= c))
  let fn (Dataset xs) c = List.count xs ~f:(fun (x, b) -> b && Float.(x < c))
  let _recall d c =
    let tp = tp d c in
    float tp /. float (tp + fn d c)
  let precision d c =
    let tp = tp d c in
    float tp /. float (tp + fp d c)
  let average_precision (Dataset xs as d) =
    if List.is_empty xs then invalid_arg "average_precision estimator undefined on empty lists" ;
    let ys = List.filter_map xs ~f:(fun (x, b) -> if b then Some x else None) in
    let n = List.length ys in
    List.fold ys ~init:0. ~f:(fun acc y -> precision d y +. acc)
    /. float n
end

module Sklearn_implementation = struct
  let average_precision (Dataset xs) =
    let y_score = List.map xs ~f:fst |> Array.of_list |> Np.Numpy.vectorf in
    let y_true = List.map xs ~f:(fun (_, b) -> Bool.to_int b) |> Array.of_list |> Np.Numpy.vectori in
    Sklearn.Metrics.average_precision_score ~y_score ~y_true ()
end

let rec binormal_generator ~n rng =
  let alpha = Gsl.Randist.flat rng ~a:0.1 ~b:1. in
  let model = Binormal_model.make alpha in
  let Dataset xs as d = Binormal_model.simulation rng ~n model in
  if List.exists xs ~f:snd then d
  else binormal_generator ~n rng

let saturated_binormal_generator ~n rng =
  let clip a b x = Float.(if x < a then a else if x > b then b else x) in
  let Dataset xs = binormal_generator ~n rng in
  Dataset (List.map xs ~f:(fun (x, b) -> clip (-0.5) 1.5 x, b))

let test_implementations ~gen ~n ~show ~equal f g () =
  let rng = Gsl.Rng.(make (default ())) in
  let rec loop i =
    if i < n then
      let x = gen rng in
      if not (equal (f x) (g x)) then (
        Alcotest.fail (sprintf "Implementations differ on %s" (show x))
      )
      else loop (i + 1)
  in
  loop 0

type dataset = Prc.dataset = Dataset of (float * bool) list
[@@deriving show]

let average_precision_test gen =
  test_implementations
    ~gen ~show:show_dataset ~n:100 ~equal:floats_are_close_enough
    auc_average_precision
    Naive_implementation.average_precision

let average_precision_sklearn_test gen =
  test_implementations
    ~gen ~show:show_dataset ~n:100 ~equal:floats_are_close_enough
    auc_average_precision
    Naive_implementation.average_precision

let average_precision_test_suite f = [
  "average_precision on binormal model 3", `Quick, f (binormal_generator ~n:3) ;
  "average_precision on binormal model 10", `Quick, f (binormal_generator ~n:10) ;
  "average_precision on binormal model 100", `Quick, f (binormal_generator ~n:100) ;
  "average_precision on saturated binormal model 3", `Quick, f (saturated_binormal_generator ~n:3) ;
  "average_precision on saturated binormal model 10", `Quick, f (saturated_binormal_generator ~n:10) ;
  "average_precision on saturated binormal model 100", `Quick, f (saturated_binormal_generator ~n:100) ;
]

let naive_implementation_suite = average_precision_test_suite average_precision_test
let sklearn_suite = average_precision_test_suite average_precision_sklearn_test

let () =
  Alcotest.run "prc" [
    "test-against-naive-implementation", naive_implementation_suite ;
    "test-against-sklearn", sklearn_suite ;
  ]


lib/prc.ml
open Core_kernel
open Gsl

let sum a b ~f =
  let rec loop i acc =
    if i > b then acc
    else loop (i + 1) (acc +. f i)
  in
  loop a 0.

let logit p = Float.log (p /. (1. -. p))

let sigmoid x =
  let exp_x = Float.exp x in
  exp_x /. (1. +. exp_x)

type dataset = Dataset of (float * bool) list

let n_pos (Dataset xs) = List.count xs ~f:(fun (_, b) -> b)

let confusion_matrix_fold (Dataset d) ~init ~f =
  let xs = Array.of_list d in
  Array.sort xs ~compare:(fun (x, _) (y, _) -> Float.compare y x) ;
  let n = Array.length xs in
  let npos = Array.count xs ~f:snd in
  let nneg = n - npos in
  let rec loop i acc ~tp ~fp ~fn ~tn =
    if i <= n then
      let score, b = xs.(i - 1) in
      let tp = if b then tp + 1 else tp
      and fp = if b then fp else fp + 1
      and fn = if b then fn - 1 else fn
      and tn = if b then tn else tn - 1 in
      let acc =
        if i = n || Float.(score <> fst xs.(i)) then
          f ~tp ~fp ~fn ~tn ~score acc
        else acc
      in
      loop (i + 1) acc ~tp ~tn ~fp ~fn
    else acc
  in
  let tp = 0 and tn = nneg and fp = 0 and fn = npos in
  loop 1 (f ~tp ~fp ~fn ~tn ~score:Float.infinity init) ~tp ~fp ~tn ~fn

type confusion_matrix = {
  tp : int ;
  tn : int ;
  fp : int ;
  fn : int ;
}
[@@deriving sexp]

type confusion_matrix_series = (float * confusion_matrix) array
[@@deriving sexp]

let confusion_matrix_series d =
  confusion_matrix_fold d ~init:[] ~f:(fun ~tp ~fp ~fn ~tn ~score acc ->
      (score, { tp ; tn ; fp ; fn }) :: acc
    )
  |> List.rev
  |> Array.of_list

let%expect_test "performance curve 1" =
  let scores = [2.1 ; 1.2 ; 5.6 ; 0. ; 5.6] in
  let labels = [true ; false ; true ; false ; false] in
  let dataset = Dataset (List.zip_exn scores labels) in
  let series = confusion_matrix_series dataset in
  print_endline (Sexp.to_string_hum (sexp_of_confusion_matrix_series series)) ;
  [%expect "
    ((INF ((tp 0) (tn 3) (fp 0) (fn 2))) (5.6 ((tp 1) (tn 2) (fp 1) (fn 1)))
     (2.1 ((tp 2) (tn 2) (fp 1) (fn 0))) (1.2 ((tp 2) (tn 1) (fp 2) (fn 0)))
     (0 ((tp 2) (tn 0) (fp 3) (fn 0))))"]

let recall ~tp ~fn = float tp /. float (tp + fn)
let precision ~tp ~fp =
  if tp + fp = 0 then 1.
  else float tp /. float (tp + fp)

let operating_points d =
  confusion_matrix_fold d ~init:[] ~f:(fun ~tp ~fp ~fn ~tn:_ ~score acc ->
      (score, recall ~tp ~fn, precision ~tp ~fp) :: acc
    )
  |> List.rev

let auc_trapezoidal_lt d =
  let points = operating_points d in
  let n, r, pmin, pmax =
    let data =
      List.group points ~break:(fun (_, r1, _) (_, r2, _) ->
          Float.(r1 <> r2)
        )
      |> List.map ~f:(function
          | [] -> assert false
          | (_, r, _) :: _ as xs -> r, List.map xs ~f:trd3
        )
      |> Array.of_list
    in
    Array.length data,
    Array.map data ~f:fst,
    Array.map data ~f:(fun (_, ps) -> List.reduce_exn ps ~f:Float.min),
    Array.map data ~f:(fun (_, ps) -> List.reduce_exn ps ~f:Float.max)
  in
  sum 0 (n - 2) ~f:(fun i -> (pmin.(i) +. pmax.(i + 1)) /. 2. *. (r.(i + 1) -. r.(i)))

let decreasing_ties_groups (Dataset xs) =
  let rec loop closed_groups current_group points =
    match points, current_group with
    | [], None -> []
    | [], Some (s, npos, nneg) ->
      (s, npos, nneg) :: closed_groups
    | (s, b) :: t, None ->
      let npos, nneg = if b then 1, 0 else 0, 1 in
      loop closed_groups (Some (s, npos, nneg)) t
    | (s', b) :: t, Some (s, npos, nneg) ->
      if Float.equal s s' then
        let npos, nneg = if b then npos + 1, nneg else npos, nneg + 1 in
        loop closed_groups (Some (s, npos, nneg)) t
      else
        let npos', nneg' = if b then 1, 0 else 0, 1 in
        loop ((s, npos, nneg) :: closed_groups) (Some (s', npos', nneg')) t
  in
  let compare = Tuple.T2.compare ~cmp1:Float.ascending ~cmp2:Bool.ascending in
  List.sort xs ~compare
  |> loop [] None

let decreasing_ties_groups_test d =
  let groups = decreasing_ties_groups d in
  Sexplib.Sexp.pp Format.std_formatter ([%sexp_of: (float * int * int) list] groups)

let%expect_test "decreasing_ties_groups" =
  let data = Dataset [0., false; 1., false; 2., true; 0., false; 1., false; 2., true; 0., true; 1., false; 2., true; 0., true] in
  decreasing_ties_groups_test data ;
  [%expect "((2 3 0)(1 0 3)(0 2 2))"]

let%expect_test "decreasing_ties_groups 2" =
  let data = Dataset [(1.0070885317, true); (0.297475057516, false); (0.831050790341, false)] in
  decreasing_ties_groups_test data ;
  [%expect "((1.0070885317 1 0)(0.831050790341 0 1)(0.297475057516 0 1))"]

let auc_average_precision (Dataset xs as d) =
  if List.is_empty xs then invalid_arg "average_precision estimator undefined on empty lists" ;
  let npos, _, sum =
    decreasing_ties_groups d
    |> List.fold ~init:(0, 0, 0.) ~f:(fun (npos, nneg, sum) (_, p, n) ->
        let npos = npos + p and nneg = nneg + n in
        let prec = float npos /. float (npos + nneg) in
        let sum = float p *. prec +. sum in
        npos, nneg, sum
      )
  in
  sum /. float npos

let logit_confidence_interval ~alpha ~theta_hat ~n_pos =
  let eta_hat = logit theta_hat in
  let tau_hat = (float n_pos *. theta_hat *. (1. -. theta_hat)) ** (-0.5) in
  let delta = tau_hat *. Cdf.gaussian_Pinv ~sigma:1. ~p:(1. -. alpha /. 2.) in
  sigmoid (eta_hat -. delta),
  sigmoid (eta_hat +. delta)

let resample_array rng xs =
  let n = Array.length xs in
  Array.map xs ~f:(fun _ ->
      let k = Gsl.Rng.uniform_int rng n in
      xs.(k)
    )

let bootstrap_confidence_interval ?(niter = 1000) ~alpha rng (Dataset xs) ~f =
  let pos, neg = Array.partition_tf ~f:snd (Array.of_list xs) in
  let resample rng =
    let xs = Array.append (resample_array rng pos) (resample_array rng neg) in
    Dataset (Array.to_list xs)
  in
  let vals = Array.init niter ~f:(fun _ -> f (resample rng)) in
  Array.sort vals ~compare:Float.compare ;
  let quantile = Gsl.Stats.quantile_from_sorted_data vals in
  quantile (alpha /. 2.), quantile (1. -. alpha /. 2.)

module Binormal_model = struct
  type t = {
    mu_pos : float ;
    sigma_pos : float ;
    mu_neg : float ;
    sigma_neg : float ;
    alpha : float ;
  }

  let make ?(mu_pos = 1.) ?(sigma_pos = 1.) ?(mu_neg = 0.) ?(sigma_neg = 1.) alpha =
    { mu_pos ; mu_neg ; sigma_pos ; sigma_neg ; alpha }

  let simulation rng ~n { mu_pos ; mu_neg ; sigma_pos ; sigma_neg ; alpha } =
    let sample = List.init n ~f:(fun _ ->
        let mu, sigma, label = match Randist.bernoulli rng ~p:alpha with
          | 0 -> mu_neg, sigma_neg, false
          | 1 -> mu_pos, sigma_pos, true
          | _ -> assert false
        in
        mu +. Randist.gaussian rng ~sigma, label
      )
    in
    Dataset sample

  let phi ~mu ~sigma x = Cdf.gaussian_P ~x:(x -. mu) ~sigma
  let inv_phi ~mu ~sigma p = Cdf.gaussian_Pinv ~p ~sigma +. mu

  let precision p recall =
    let phi_neg = phi ~mu:p.mu_neg ~sigma:p.sigma_neg in
    let inv_phi_pos = inv_phi ~mu:p.mu_pos ~sigma:p.sigma_pos in
    let alpha = p.alpha in
    alpha *. recall
    /.
    (alpha *. recall
     +.
     (1. -. alpha) *. (1. -. phi_neg (inv_phi_pos (1. -. recall))))

  let curve ?(n = 100) p =
    let max_i = float (n - 1) in
    Array.init n ~f:(fun i ->
        let r = float i /. max_i in
        r, precision p r
      )

  let estimate (Dataset xs) =
    let select label =
      List.filter_map xs ~f:(fun (x, b) ->
          if Bool.(b = label) then Some x else None
        )
      |> Array.of_list
    in
    let x = select false in
    let y = select true in
    let len t = float (Array.length t) in
    Stats.{
      mu_pos = mean y ;
      mu_neg = mean x ;
      sigma_pos = sd y ;
      sigma_neg = sd x ;
      alpha = len y /. (len x +. len y) ;
    }

  let auc p =
    let n = 1_000 in
    let x = Array.init n ~f:(fun i -> float (i + 1) /. float n) in
    let y = Array.map x ~f:(precision p) in
    Stats.mean y
end


demo/prc_demo.ml
open Core_kernel
open Prc
open Gsl

module Plot = struct
  type t = {
    x : float array ;
    y : float array ;
    col : string option ;
    ty : [`Lines of int | `Points of int] ;
    label : string option ;
  }

  let lines ?col ?(lwd = 1) ?label points = {
    x = Array.map points ~f:fst ;
    y = Array.map points ~f:snd ;
    col ; ty = `Lines lwd ; label ;
  }

  let points ?col ?(pch = 19) ?label points = {
    x = Array.map points ~f:fst ;
    y = Array.map points ~f:snd ;
    col ; ty = `Points pch ; label ;
  }

  let plot_curve { x ; y ; col ; ty ; _ } =
    match ty with
    | `Lines lwd ->
      OCamlR_graphics.lines ~lwd ?col ~x ~y ()
    | `Points pch ->
      OCamlR_graphics.points ~pch ?col ~x ~y ()

  let recall_precision_plot ?main curves =
    OCamlR_graphics.plot
      ~plot_type:`Lines ?main
      ~xlab:"Recall" ~ylab:"Precision"
      ~xlim:(0., 1.) ~ylim:(0., 1.) ~x:[||] ~y:[||] () ;
    List.(iter (rev curves)) ~f:plot_curve
end

let recall_precision_curve ?col ?lwd ?label dataset =
  Prc.operating_points dataset
  |> List.map ~f:(fun (_, recall, precision) -> recall, precision)
  |> Array.of_list
  |> Plot.lines ?col ?lwd ?label

let curve_variability ?(sample_size = 1000) () =
  let rng = Rng.(make (default ())) in
  let alpha = 0.1 in
  let p = Binormal_model.make alpha in
  let samples = List.init 20 ~f:(fun _ -> Binormal_model.simulation ~n:sample_size rng p) in
  let empirical_curves =
    List.map samples ~f:(recall_precision_curve ~col:"grey")
  in
  let true_curve =
    Binormal_model.curve ~n:100 p
    |> Plot.lines ~col:"red" ~label:"True RP curve"
  in
  Plot.recall_precision_plot (true_curve :: empirical_curves)

let estimator_sampling_distribution ?(sigma = 1.) ?(alpha = 0.1) ?(sample_size = 30) () =
  let open OCamlR_base in
  let rng = Rng.(make (default ())) in
  let p = Binormal_model.make ~sigma_pos:sigma ~sigma_neg:sigma alpha in
  let true_auc = Binormal_model.auc p in
  let samples = List.init 1_000 ~f:(fun _ -> Binormal_model.simulation ~n:sample_size rng p) in
  let compute f =
    List.map samples ~f
    |> Numeric.of_list
    |> Numeric.to_sexp
  in
  let trapezoidal_lt = compute auc_trapezoidal_lt in
  let average_precision = compute auc_average_precision in
  let binormal = compute (fun sample -> Binormal_model.auc (Binormal_model.estimate sample)) in
  let l = List_.create [
      Some "binormal", binormal ;
      Some "trapezoidal", trapezoidal_lt ;
      Some "average_precision", average_precision ;
    ]
  in
  OCamlR_graphics.list_boxplot l ;
  OCamlR_graphics.abline ~h:true_auc ~lty:`dashed ~col:"red" ~lwd:2 ()

let confidence_interval_coverage meth sample_size =
  let rng = Rng.(make (default ())) in
  let model = Binormal_model.make 0.1 in
  let alpha = 0.05 in
  let true_auc = Binormal_model.auc model in
  let confidence_interval dataset = match meth with
    | `Logit ->
      let auc_estimate = auc_average_precision dataset in
      let n_pos = n_pos dataset in
      logit_confidence_interval ~alpha ~theta_hat:auc_estimate ~n_pos
    | `Bootstrap ->
      bootstrap_confidence_interval ~alpha Rng.(make (default ())) dataset ~f:auc_average_precision
  in
  let iteration _ =
    let dataset = Binormal_model.simulation rng ~n:sample_size model in
    let (lb, ub) = confidence_interval dataset in
    Float.(lb < true_auc && true_auc < ub)
  in
  Array.init 100 ~f:iteration
  |> Array.count ~f:Fn.id
